include/net/dsa.h

```diff
@@ -0,0 +1,174 @@
+/** @file
+ @brief DSA definitions and handlers
+
+ */
+
+/*
+ * Copyright (c) 2020 DENX Software Engineering GmbH
+ *               Lukasz Majewski <lukma@denx.de>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_DSA_H_
+#define ZEPHYR_INCLUDE_NET_DSA_H_
+
+#include <device.h>
+#include <net/net_if.h>
+
+/**
+ * @brief LLDP definitions and helpers
+ * @defgroup lldp Link Layer Discovery Protocol definitions and helpers
+ * @ingroup networking
+ * @{
+ */
+
+#define NET_DSA_PORT_MAX_COUNT 8
+
+#define DSA_NET_DEVICE_INIT(dev_name, drv_name, init_fn, pm_control_fn,	\
+			    data, cfg_info, prio, api, mtu, port_num)	  \
+	DEVICE_DEFINE(dev_name, drv_name, init_fn, pm_control_fn, data,	\
+		      cfg_info, POST_KERNEL, prio, api);		\
+	NET_L2_DATA_INIT(dev_name, 0, NET_L2_GET_CTX_TYPE(ETHERNET_L2)); \
+	NET_IF_INIT(dev_name, 0, ETHERNET_L2, mtu, port_num)
+
+#define DSA_STATUS_TICK_MS K_MSEC(1000)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief DSA function to adjust packet
+ *
+ * This is a function for adjusting packets passed from slave DSA interface to
+ * master (e.g. implement tail tagging).
+ *
+ * @param dev Device
+ * @param pkt Network packet
+ *
+ * Returns:
+ *  - Pointer to (modified) net_pkt structure
+ */
+struct net_pkt *dsa_xmit_pkt(struct device *dev, struct net_pkt *pkt);
```

Comment:
You will need to make this public API (or at least part of it, if I see your sample) accessible through net_mgmt, instead of exposing such functions. Then it will by default be ready for userland access (via its socket interface).

Reply:
I do have  more questions regarding this issue - i.e. with using the net_mgmt callbacks.
I've created an issue to discuss this:
https://github.com/zephyrproject-rtos/zephyr/issues/25187